### ISSUE TITLE
Panicky, non-validating conversions from/to protobuf types

### ISSUE THREAD

**Issue by mzabaluev** [informalsystems/tendermint-rs#1037]
After #1022 has landed, there is a number of conversions between protobuf types and so-called domain types that do not validate their input, are prone to panic when the input does not conform to expectations on the destination value, or produce a protobuf message struct that falls afoul of the [specification](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf) on well-known protobuf types.

Here's why this is problematic for protobuf ⭢ domain type:
As the most likely origin of protobuf DTO values is by having been parsed from an incoming message by prost without any validation as to its contents, this is an easy DoS bomb under application developers who would make use of these conversions. Please consider using fallible conversions when dealing with untrusted values originating from the network.

Here's why this is problematic for domain type ⭢ protobuf:
Aside from the low-level tendermint-proto crate, tendermint-rs should never produce protocol messages that do not comply with the specification. The alternative of panicking makes for poor API usability. The tendermint API should provide true domain types where `From`-type conversion to their protobuf DTOs is a total function always resulting in valid messages.

_Originally posted by @mzabaluev in https://github.com/informalsystems/tendermint-rs/pull/1030#discussion_r758939392_

**Comment by tony-iqlusion** [comment]
FWIW, in `cosmrs` I used a trait with an associated type to enforce conversion bounds between "domain types" and their associated prost-generated protobufs:

https://docs.rs/cosmrs/0.3.0/cosmrs/tx/trait.Msg.html

```rust
pub trait Msg: Clone + Sized + TryFrom<Self::Proto, Error = ErrorReport> + Into<Self::Proto> {
    type Proto: MsgProto;
    [...]
}
```

This is somewhat specialized to the `tx.Msg` use case, but with the elided portion aside it works for any "domain type" with an associated protobuf type.

The bounds ensure that there's always a `TryFrom` conversion available from the associated Protobuf type, as well as an `Into` conversion to convert to the domain type.

**Comment by thanethomson** [comment]
Would the approach outlined in #661 be of use here?

**Comment by mzabaluev** [comment]
> Would the approach outlined in #661 be of use here?

Yes, essentially this issue is tracking transgressions against that approach.
